### PR TITLE
Deployment Scripts cmdlets

### PR DESCRIPTION
**Pull request by filizt**
## Description

This PR adds deployment script cmdlets to the Az.Resources module.

Link to the cmdlet design review: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/400
 
## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/GetAzDeploymentScriptLog.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        [ValidateNotNullOrEmpty]
+        public PsDeploymentScript DeploymentScriptInputObject { get; set; }
+
+        [Parameter(Position = 2, ParameterSetName = GetDeploymentScriptLogByName, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The directory path to save deployment script log.")]
+        [Parameter(Position = 1, ParameterSetName = GetDeploymentScriptLogByResourceId, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The directory path to save deployment script log.")]
+        [Parameter(Position = 1, ParameterSetName = GetDeploymentScriptLogByInputObject, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The directory path to save deployment script log.")]
+        [ValidateNotNullOrEmpty]
+        public string OutputPath { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "When set to true, execution will not ask for a confirmation for the operation.")]
```
Comment (isra-fel) [inline src/Resources/ResourceManager/Implementation/DeploymentScripts/GetAzDeploymentScriptLog.cs]:
```suggestion
        [Parameter(Mandatory = false, HelpMessage = "Overwrite the given file if it exists.")]
```

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/GetAzDeploymentScriptLog.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+                if (!string.IsNullOrEmpty(OutputPath) && !string.IsNullOrEmpty(deploymentScriptLog?.Log))
+                {
+                    var outputPathWithFileName = Path.Combine(GetValidatedFolderPath(OutputPath), $"{deploymentScriptLog.DeploymentScriptName}.txt");
+
+                    // if a file with the same name exists, let's ask if we can replace it. If it doesn't exist, we'll create it right away.
+                    this.ConfirmAction(
+                        this.Force || !AzureSession.Instance.DataStore.FileExists(outputPathWithFileName),
+                        string.Format(
+                            "A file with the name {0} already exists in directory {1}. This operation will replace the existing file. Would you like to continue?",
```
Comment (isra-fel) [inline src/Resources/ResourceManager/Implementation/DeploymentScripts/GetAzDeploymentScriptLog.cs]: Please use resource string for customer-facing messages.

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/RemoveAzDeploymentScript.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        [Parameter(Position = 1, ParameterSetName = RemoveDeploymentScriptByName, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The name of the resource group.")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Alias("ResourceId")]
+        [Parameter(Position = 0, ParameterSetName = RemoveDeploymentScriptByResourceId, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The fully qualified resource Id of the deployment script. Example: /subscriptions/{subId}/resourceGroups/{rgName}/providers/Microsoft.Resources/deploymentScripts/{deploymentScriptName}")]
+        [ValidateNotNullOrEmpty]
+        public string Id { get; set; }
```
Comment (isra-fel) [inline src/Resources/ResourceManager/Implementation/DeploymentScripts/RemoveAzDeploymentScript.cs]:
```suggestion
        [Alias("Id")]
        [Parameter(Position = 0, ParameterSetName = RemoveDeploymentScriptByResourceId, Mandatory = true, ValueFromPipelineByPropertyName = true,
            HelpMessage = "The fully qualified resource Id of the deployment script. Example: /subscriptions/{subId}/resourceGroups/{rgName}/providers/Microsoft.Resources/deploymentScripts/{deploymentScriptName}")]
        [ValidateNotNullOrEmpty]
        public string ResourceId { get; set; }
```

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+        [Fact()]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void TestDeploymentScriptGetForCli()
+        {
+            TestRunner.RunTestScript("Test-GetDeploymentScript-Cli");
+        }
+
+        [Fact(Skip = "Need to wait until the Resources .NET SDK is updated to 3.3.0 in the PS module.")]
```
Comment (isra-fel) [inline src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs]: When do you plan to update the SDK

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+        [Fact()]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void TestDeploymentScriptLogGetForCli()
+        {
+            TestRunner.RunTestScript("Test-GetDeploymentScriptLog-Cli");
+        }
+
+        [Fact(Skip = "Saving a file operation fails. To-Do Ask test infra team. ")]
```
Comment (isra-fel) [inline src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs]: What's about the failure?

File: src/Resources/Resources/help/Get-AzDeploymentScriptLog.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+Aliases:
+
+Required: True
+Position: 0
+Default value: None
+Accept pipeline input: True (ByPropertyName)
+Accept wildcard characters: False
+```
+
+### -Force
```
[the remark this answers is not in the document]
Comment (filizt) [reply src/Resources/Resources/help/Get-AzDeploymentScriptLog.md]: @isra-fel the -Force parameter was for saving the log locally. I'll move all this to new Save-AzDeploymentScriptLog cmdlet.

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/GetAzDeploymentScript.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+        [Parameter(Position = 1, ParameterSetName = GetDeploymentScriptByName, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The name of the deployment script")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Alias("ResourceId")]
+        [Parameter(Position = 0, ParameterSetName = GetDeploymentScriptByResourceId, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The fully qualified resource Id of the deployment script. Example: /subscriptions/{subId}/resourceGroups/{rgName}/providers/Microsoft.Resources/deploymentScripts/{deploymentScriptName}")]
+        [ValidateNotNullOrEmpty]
+        public string Id { get; set; }
```
[the remark this answers is not in the document]
Comment (filizt) [reply src/Resources/ResourceManager/Implementation/DeploymentScripts/GetAzDeploymentScript.cs]: We're consistent with the Deployment(e.g: GetAzureSubscriptionDeployment) cmdlets with this usage. Let me know if you have a strong opinion on this. 

---
In reply to: [390125815](https://github.com/Azure/azure-powershell/pull/11266#discussion_r390125815) [](ancestors = 390125815)

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/RemoveAzDeploymentScript.cs (hunk 3 above)
Comment (filizt) [reply src/Resources/ResourceManager/Implementation/DeploymentScripts/RemoveAzDeploymentScript.cs]: Same with the other comment - We're consistent with the Deployment(e.g: GetAzureSubscriptionDeployment) cmdlets with this usage. Let me know if you have a strong opinion on this. 

---
In reply to: [390159975](https://github.com/Azure/azure-powershell/pull/11266#discussion_r390159975) [](ancestors = 390159975)

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs (hunk 4 above)
Comment (filizt) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs]: There will be a separate PR after we have this version of the cmdlets checked in. 

---
In reply to: [390161237](https://github.com/Azure/azure-powershell/pull/11266#discussion_r390161237) [](ancestors = 390161237)

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs (hunk 5 above)
Comment (filizt) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs]: The execution will not save the log file locally during. I might be missing something here. I'll reach out to you offline for this.

---
In reply to: [390161450](https://github.com/Azure/azure-powershell/pull/11266#discussion_r390161450) [](ancestors = 390161450)

**Comment by filizt** [comment]
@ I created Save-AzDeploymentScriptLog cmdlet.

---
In reply to: [371716364](https://github.com/Azure/azure-powershell/pull/11266#issuecomment-371716364) [](ancestors = 371716364)

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs (hunk 5 above)
Comment (filizt) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScriptsTests.cs]: I have joined in the PS office hours today about this. I was told to implement in memory datastore mock and test against that. I'll work on that and submit this test.

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/SaveAzDeploymentScriptLog.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            HelpMessage = "Forces the overwrite of the existing file.")]
+        public SwitchParameter Force { get; set; }
+
+        #endregion
+
+        #region Cmdlet Overrides
+
+        public override void ExecuteCmdlet()
+        {
+           // base.OnProcessRecord();
```
Comment (isra-fel) [inline src/Resources/ResourceManager/Implementation/DeploymentScripts/SaveAzDeploymentScriptLog.cs]: Please remove commented code if it's not necessary 😀

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/SaveAzDeploymentScriptLog.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            HelpMessage = "The deployment script PowerShell object.")]
+        [ValidateNotNullOrEmpty]
+        public PsDeploymentScript DeploymentScriptInputObject { get; set; }
+
+        [Parameter(Position = 2, ParameterSetName = SaveDeploymentScriptLogByName, Mandatory = true,
+            ValueFromPipelineByPropertyName = true, HelpMessage = "The directory path to save deployment script log.")]
+        [Parameter(Position = 1, ParameterSetName = SaveDeploymentScriptLogByResourceId, Mandatory = true,
+            ValueFromPipelineByPropertyName = true, HelpMessage = "The directory path to save deployment script log.")]
+        [Parameter(Position = 1, ParameterSetName = SaveDeploymentScriptLogByInputObject, Mandatory = true,
+            ValueFromPipelineByPropertyName = true, HelpMessage = "The directory path to save deployment script log.")]
```
Comment (isra-fel) [inline src/Resources/ResourceManager/Implementation/DeploymentScripts/SaveAzDeploymentScriptLog.cs]: There isn't gonna be a piping scenario that the output of the first cmdlet contains a property named "OutputPath", so it's not necessary to use `ValueFromPipelineByPropertyName`.
```suggestion
        [Parameter(Position = 2, ParameterSetName = SaveDeploymentScriptLogByName, Mandatory = true,
            HelpMessage = "The directory path to save deployment script log.")]
        [Parameter(Position = 1, ParameterSetName = SaveDeploymentScriptLogByResourceId, Mandatory = true,
            HelpMessage = "The directory path to save deployment script log.")]
        [Parameter(Position = 1, ParameterSetName = SaveDeploymentScriptLogByInputObject, Mandatory = true,
            HelpMessage = "The directory path to save deployment script log.")]
```

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/RemoveAzDeploymentScript.cs (hunk 3 above)
Comment (isra-fel) [reply src/Resources/ResourceManager/Implementation/DeploymentScripts/RemoveAzDeploymentScript.cs]: I'm OK with that!

File: src/Resources/ResourceManager/Implementation/DeploymentScripts/SaveAzDeploymentScriptLog.cs (hunk 9 above)
Comment (filizt) [reply src/Resources/ResourceManager/Implementation/DeploymentScripts/SaveAzDeploymentScriptLog.cs]: Updated.

---
In reply to: [392791889](https://github.com/Azure/azure-powershell/pull/11266#discussion_r392791889) [](ancestors = 392791889)

**Comment by filizt** [comment]
> Hi @filizt the PR looks great. Please fix two minor problems, resolve the conflict, and regenerate help docs. Thank you.

@isra-fel Addressed your comments, resolved conflict, regenerated helps docs, re-recorded the tests and added a new test. I also updated the change logs. Let me know if I'm missing anything.

**Comment by filizt** [comment]
@isra-fel I'm looking at the CI failures. The erros for .md files say "Online version in the header of the file is incorrect. The corresponding URL should be t." Where would I find the online help URL?

**Pull request merged**